### PR TITLE
[Perf] Split-reduction batch-invariant log_softmax for small-batch decode

### PR DESCRIPTION
**Pull request by shivasathishs-rp**
## Purpose

Part of the batch-invariant performance work tracked in #27433 ("Optimize the batch invariant performance").

The batch-invariant `log_softmax` (`vllm/model_executor/layers/batch_invariant.py`) launches **one Triton program per row** (`grid = (n_rows,)`), and each program scans the full vocab (128k–152k) in three serial passes with a fixed `BLOCK_SIZE=1024`. In decode, `n_rows` equals the batch size, so when it is below the SM count (~108 on A100) most SMs sit idle while a handful of programs serially reduce a very wide row. This kernel is registered for **all CUDA archs** (not just SM80) and sits on the always-on logprob path — `Sampler.compute_logprobs` → `logits.log_softmax(dim=-1, dtype=torch.float32)` — which dominates batch-invariant RL rollouts, the feature's headline use case.

## Change

Split each row's reduction across the column (vocab) dimension:

- `_log_softmax_stats_kernel` over `grid = (n_rows, n_splits)` computes per-split `(max, sum_exp)` partials.
- `_log_softmax_finalize_kernel` combines the partials in fixed split order (rescaling each split's `sum_exp` to the row's global max) and writes the output.

Small-batch decode now occupies many SMs instead of one program per row.

## Batch invariance is preserved

`n_splits` is derived **only** from the vocab width (`n_cols`), never from `n_rows`, so a row is partitioned and reduced in an identical, batch-independent order regardless of how many rows share the launch. `max` is order-free, and each split's `sum_exp` is rescaled to the global max before a fixed-order combine, so the reduction order depends solely on `n_splits`. This is the same discipline used elsewhere in the batch-invariant kernels (e.g. pinning `BLOCK_SIZE_K` in the matmul path).

Verified on A100 under `VLLM_BATCH_INVARIANT=1` that a row's output is **bitwise identical** (`torch.equal`) whether processed alone (M=1) or embedded in batches up to 512, across vocab sizes {2048, 32000, 151936} and bf16/fp32, and across batch sizes {1, 2, 8, 33, 64, 128, 257}. Output is also bitwise identical to the previous kernel (`max|old − new| = 0.0`).

## Test plan and results (A100-40GB)

```
VLLM_BATCH_INVARIANT=1 pytest tests/v1/determinism/test_log_softmax_batch_invariant.py -v
# 13 passed
```

New tests in `tests/v1/determinism/test_log_softmax_batch_invariant.py`:
- `test_log_softmax_batch_invariance` — a row is bitwise identical alone vs. embedded in a batch of 512, across vocab sizes and dtypes.
- `test_log_softmax_matches_reference` — split reduction matches `torch.log_softmax`.
- `test_log_softmax_invariant_across_batch_launch_boundaries` — a fixed row is stable across batch sizes straddling SM occupancy.

Performance (vocab=151936, bf16, A100), old (one program/row) vs new (split):

| M (rows) | old µs | new µs | speedup |
|---:|---:|---:|---:|
| 1 | 187.6 | 81.1 | 2.31x |
| 8 | 191.9 | 80.9 | 2.37x |
| 64 | 285.0 | 122.2 | 2.33x |
| 256 | 420.9 | 281.9 | 1.49x |
| 1024 | 938.8 | 991.3 | 0.95x |

The win is in the small-M decode regime the batch-invariant logprob path lives in. There is a ~5% regression at M=1024, where rows alone already fill the SMs; this is an inherent consequence of keeping `n_splits` a function of vocab width only (making it depend on `n_rows` would break batch invariance). Large-M logprob passes are rare on this path, so the tradeoff favors the decode win.

## Why this is not duplicating an existing PR

No open or merged PR touches the batch-invariant `log_softmax` kernel. I searched open PRs for "batch invariant log_softmax", "log_softmax kernel", "split reduction log softmax", and PRs touching `batch_invariant.py` — the only hits are #45819 (an unrelated GDN_ATTN feature) and #35885 (an unrelated logprobs-mode bugfix). The tracker's optimized items (bmm #29345, fused RMSNorm #40413, cutlass fp8 #40408) are all merged and do not cover log_softmax.

## Model evaluation

Output is bitwise identical to the previous kernel (`max|old − new| = 0.0`) and matches `torch.log_softmax` within tolerance, so model quality is unchanged. This only alters intra-kernel parallelization; the reduction result is unchanged.

## AI assistance disclosure

AI assistance was used to help localize the opportunity, draft the kernel and tests, and run the A100 validation above. I reviewed every changed line and ran the tests.

### CODE REVIEW

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work!

Could you also test E2E performance using `vllm bench ...`?

Also, please test using Hopper/Blackwell as A100 is not the main target for optimization.

**Comment by shivasathishs-rp** [comment]
Added an end-to-end check on top of the kernel tests, since this sits on the sampler logprob path. Ran Qwen2.5-7B on an A100 under `VLLM_BATCH_INVARIANT=1` (eager, greedy) and compared a fixed sequence's per-step logprobs decoded alone vs. embedded in a batch of 64: bit-identical (0/96 positions differ). Also compared against the old one-program-per-row kernel over both batch sizes and it's a numerical no-op there too (max abs diff 0.0). So the split doesn't perturb anything downstream, as expected.
